### PR TITLE
[Wait for #2710][Layer] add "subtract layer" @open sesame 11/06 18:29

### PR DESCRIPTION
**Pull request by baek2sm**
- added "subtract layer"

**Self evaluation:**
1. Build test:   [X]Passed [ ]Failed [ ]Skipped
2. Run test:     [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Seungbaek Hong <sb92.hong@samsung.com>

### CODE REVIEW

File: test/unittest/models/unittest_models.cpp [hunk 1]
```diff
@@ -892,9 +930,9 @@ GTEST_PARAMETER_TEST(
                  "multi_head_attention_float_attn_mask",
                  ModelTestOption::ALL_V2),
     /** @todo:change model if bool type tensor is supported */
-    mkModelTc_V2(makeMultiHeadAttention_float_attn_mask,
-                 "multi_head_attention_pseudo_bool_attn_mask",
-                 ModelTestOption::ALL_V2),
+    //  mkModelTc_V2(makeMultiHeadAttention_float_attn_mask,
```
Comment (EunjuYang) [inline test/unittest/models/unittest_models.cpp:933]: Could you add a comment explaining why this part is commented out?

**Review by SeoHyungjun — APPROVED** [review]
The add and sub layers have almost identical code.
How about creating a basic operation class?
It would be simple if the add and sub inherited from the basic operation class.
Multiplication and division also seem to have the same structure.

**Review by skykongkong8 — APPROVED** [review]
LGTM

**Review by djeong20 — APPROVED** [review]
Nice work! One minor suggestion is to rename it SubtractLayer (SubLayer sounds like a subunit of a layer). What do you think?

File: test/unittest/models/unittest_models.cpp (hunk 1 above)
Comment (baek2sm) [reply test/unittest/models/unittest_models.cpp:933]: sure, thanks :)

**Comment by baek2sm** [comment]
> The add and sub layers have almost identical code. How about creating a basic operation class? It would be simple if the add and sub inherited from the basic operation class. Multiplication and division also seem to have the same structure.

That sounds good. Let's think together to find an improved structure.

**Comment by baek2sm** [comment]
> Nice work! One minor suggestion is to rename it SubtractLayer (SubLayer sounds like a subunit of a layer). What do you think?

That's a good idea. I'll rename it! (and the other, too)

**Comment by jijoongmoon** [comment]
> The add and sub layers have almost identical code. How about creating a basic operation class? It would be simple if the add and sub inherited from the basic operation class. Multiplication and division also seem to have the same structure.

I also think same way. How about unary and binary layer.

File: nntrainer/layers/add_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    Tensor hidden_step = hidden_.getSharedDataTensor(
+      hidden_step_dim, b * hidden_dim.getFeatureLen(), true);
+
+    const Tensor &input0 = context.getInput(0);
+    const Tensor &input1 = context.getInput(1);
+
+    TensorDim input_dim = input0.getDim();
+    TensorDim input_step_dim = input_dim;
+    input_step_dim.batch(1);
+    input_step_dim.height(to - from);
```
Comment (EunjuYang) [inline nntrainer/layers/add_layer.cpp:66]: Similar with the comment I left for the add layer.
Do we need to put these codes here, inside of the batch loop?
What about to move them to the outside of the batch loop?

File: nntrainer/layers/subtract_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    Tensor hidden_step = hidden_.getSharedDataTensor(
+      hidden_step_dim, b * hidden_dim.getFeatureLen(), true);
+
+    const Tensor &input0 = context.getInput(0);
+    const Tensor &input1 = context.getInput(1);
+
+    TensorDim input_dim = input0.getDim();
+    TensorDim input_step_dim = input_dim;
+    input_step_dim.batch(1);
+    input_step_dim.height(to - from);
```
Comment (EunjuYang) [inline nntrainer/layers/subtract_layer.cpp:66]: Similar with the comment I left for the add layer.
Do we need to put these codes here, inside of the batch loop?
What about to move them to the outside of the batch loop?

Comment (baek2sm) [reply nntrainer/layers/subtract_layer.cpp:66]: It seems a good point. I'll apply it. Thank you!

**Comment by SeoHyungjun** [comment]
> > The add and sub layers have almost identical code. How about creating a basic operation class? It would be simple if the add and sub inherited from the basic operation class. Multiplication and division also seem to have the same structure.
> 
> That sounds good. Let's think together to find an improved structure.

How about creating the operation layer like this?


```cpp
// operation_layer.h

#pragma once    

#include <tensor.h>    

namespace nntrainer {    

template <typename OP>    
class OperationLayer {    
public:    
  virtual ~OperationLayer() = default;    

protected:    
  OperationLayer() = default;    

  virtual void finalize(InitLayerContext &context) = 0;    

  virtual void forwarding(RunLayerContext &context, bool training) = 0;    

  virtual void incremental_forwarding(RunLayerContext &context,    
                                      unsigned int from, unsigned int to,    
                                      bool training) = 0;    

  virtual void calcDerivative(RunLayerContext &context) = 0;    

  virtual void setProperty(const std::vector<std::string> &values) = 0;    

  
void commonForwarding(RunLayerContext &context, bool training,    
                       const Tensor &input0, const Tensor &input1,    
                       Tensor &hidden_);    

// Implementing functions for common codes
};    

template <>    
struct OperationLayer<ADDITION> {    
  static constexpr const char *type = "add";    
};    

template <>    
struct OperationLayer<SUBTRACTION> {    
  static constexpr const char *type = "sub";    
};    

template <>    
struct OperationLayer<MULTIPLICATION> {    
  static constexpr const char *type = "mul";    
};    

template <>    
struct OperationLayer<DIVISION> {    
  static constexpr const char *type = "div";    
};    

} // namespace nntrainer    
```

```cpp
// operation_layer.cpp

template <typename OP>  
void OperationLayer<OP>::commonForwarding(RunLayerContext& context, bool training,  
                                          const Tensor& input0, const Tensor& input1, Tensor& hidden_) {  
  switch (OP::type[0]) {  
    case 'a':  
      input0.add(input1, hidden_);  
      break;  
    case 's':  
      input0.subtract(input1, hidden_);  
      break;  
    case 'm':  
      input0.multiply(input1, hidden_);  
      break;  
    case 'd':  
      input0.divide(input1, hidden_);  
      break;  
    default:  
      throw std::runtime_error("Invalid operator type");  
  }  
}  

// implementation of other common function

```

**Comment by SeoHyungjun** [comment]
I think we can create an add layer like this using the operation layer.

```cpp
// add_layer.h
#pragma once  

#include "operation_layer.h"  

namespace nntrainer {  

class AddLayer : public OperationLayer<ADDITION> {  
public:  
  using OperationLayer<ADDITION>::OperationLayer;  

  void finalize(InitLayerContext &context) override;  

  void forwarding(RunLayerContext &context, bool training) override;  

  void incremental_forwarding(RunLayerContext &context, unsigned int from,  
                              unsigned int to, bool training) override;  

  void calcDerivative(RunLayerContext &context) override;  

  void setProperty(const std::vector<std::string> &values) override;  
};  

} // namespace nntrainer  
```

```cpp
//add_layer.cpp

#include "add_layer.h"  

namespace nntrainer {  

void AddLayer::finalize(InitLayerContext &context) {  
  // Add Layer specific initialization code goes here...  
}  

void AddLayer::forwarding(RunLayerContext &context, bool training) {  
  commonForwarding(context, training, input0, input1, hidden_);  
}  

void AddLayer::incremental_forwarding(RunLayerContext &context, unsigned int from,  
                                      unsigned int to, bool training) {  
  // Add Layer specific incremental forward propagation code goes here...  
}  

void AddLayer::calcDerivative(RunLayerContext &context) {  
  // Add Layer specific backpropagation code goes here...  
}  

void AddLayer::setProperty(const std::vector<std::string> &values) {  
  // Add Layer specific property setting code goes here...  
}  

} // namespace nntrainer 
```

There is a lot of overlap in the currently implemented code.
It seems that all other functions can be expressed simply, just like using commonForwarding in the forwarding function.

If each function requires completely different implementation codes, it may be possible to implement its own code rather than a common function in each class.

**Comment by SeoHyungjun** [comment]
> Great suggestion @SeoHyungjun ! Upon his suggestion, what about using enum for operation type? If more operations are added, the switch clause may not work only with the first character. Also it can improve readability :)

That's a good point. I changed the condition you mentioned from a single letter to a word.

**Comment by SeoHyungjun** [comment]
> Great suggestion @SeoHyungjun ! Upon his suggestion, what about using enum for operation type? If more operations are added, the switch clause may not work only with the first character. Also it can improve readability :)

Good point, @EunjuYang!!
As you mentioned, using enums would be beneficial for readability and maintenance purposes!

I have written down the code with added enums below.

```cpp
// operation_layer.h  

#pragma once      

#include <tensor.h>      

namespace nntrainer {      

enum class OpType {  
  ADDITION,  
  SUBTRACTION,  
  MULTIPLICATION,  
  DIVISION  
};  

template <OpType op_type>      
class OperationLayer {      
public:      
  virtual ~OperationLayer() = default;      

protected:      
  OperationLayer() = default;      

  virtual void finalize(InitLayerContext &context) = 0;      

  virtual void forwarding(RunLayerContext &context, bool training) = 0;      

  virtual void incremental_forwarding(RunLayerContext &context,      
                                      unsigned int from, unsigned int to,      
                                      bool training) = 0;      

  virtual void calcDerivative(RunLayerContext &context) = 0;      

  virtual void setProperty(const std::vector<std::string> &values) = 0;      

    
void commonForwarding(RunLayerContext &context, bool training,      
                       const Tensor &input0, const Tensor &input1,      
                       Tensor &hidden_);      

// Implementing functions for common codes  
};      

template <>      
struct OperationLayer<OpType::ADDITION> {      
  static constexpr const char *type = "add";      
};      

template <>      
struct OperationLayer<OpType::SUBTRACTION> {      
  static constexpr const char *type = "sub";      
};      

template <>      
struct OperationLayer<OpType::MULTIPLICATION> {      
  static constexpr const char *type = "mul";      
};      

template <>      
struct OperationLayer<OpType::DIVISION> {      
  static constexpr const char *type = "div";      
};      

} // namespace nntrainer      
```

```cpp
// operation_layer.cpp  

template <OpType op_type>    
void OperationLayer<op_type>::commonForwarding(RunLayerContext& context, bool training,    
                                              const Tensor& input0, const Tensor& input1, Tensor& hidden_) {    
  switch (op_type) {    
    case OpType::ADDITION:    
      input0.add(input1, hidden_);    
      break;    
    case OpType::SUBTRACTION:    
      input0.subtract(input1, hidden_);    
      break;    
    case OpType::MULTIPLICATION:    
      input0.multiply(input1, hidden_);    
      break;    
    case OpType::DIVISION:   
      input0.divide(input1, hidden_);    
      break;    
    default:    
      throw std::runtime_error("Invalid operator type");    
  }    
}    

// implementation of other common function  
```

```cpp
// add_layer.h  

#pragma once    

#include "operation_layer.h"    

namespace nntrainer {    

class AddLayer : public OperationLayer<OpType::ADDITION> {    
public:    
  using OperationLayer<OpType::ADDITION>::OperationLayer;    

  void finalize(InitLayerContext &context) override;    

  void forwarding(RunLayerContext &context, bool training) override;    

  void incremental_forwarding(RunLayerContext &context, unsigned int from,    
                              unsigned int to, bool training) override;    

  void calcDerivative(RunLayerContext &context) override;    

  void setProperty(const std::vector<std::string> &values) override;    
};    

} // namespace nntrainer     
```

```cpp
//add_layer.cpp  

#include "add_layer.h"    

namespace nntrainer {    

void AddLayer::finalize(InitLayerContext &context) {    
  // Add Layer specific initialization code goes here...    
}    

void AddLayer::forwarding(RunLayerContext &context, bool training) {    
  commonForwarding(context, training, input0, input1, hidden_);    
}    

void AddLayer::incremental_forwarding(RunLayerContext &context, unsigned int from,    
                                      unsigned int to, bool training) {    
  // Add Layer specific incremental forward propagation code goes here...    
}    

void AddLayer::calcDerivative(RunLayerContext &context) {    
  // Add Layer specific backpropagation code goes here...    
}    

void AddLayer::setProperty(const std::vector<std::string> &values) {    
  // Add Layer specific property setting code goes here...    
}    

} // namespace nntrainer   
```

**Comment by baek2sm** [comment]
@SeoHyungjun It's a good suggestion. I'll try to reflect on it. Thank you!

**Comment by baek2sm** [comment]
I've completed the revisions after incorporating feedbacks. Thanks a lot!

**Review by djeong20 — APPROVED** [review]
nice work!

**Pull request merged**